File: 卡码/代码随想录/1/06-二叉树/Java/binaryTree/src/main/java/com/kurumi/binarytree/_01_二叉树遍历方式/LeetCode_0515.java
//515. 在每个树行中找最大值
package com.kurumi.binarytree._01_二叉树遍历方式;

import com.kurumi.binarytree.TreeNode;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class LeetCode_0515 {

}
class Solution {
  public List<Integer> largestValues(TreeNode root) {
    List<Integer> res=new ArrayList<>();
    Queue<TreeNode> q=new LinkedList<>();
    if(root!=null) q.offer(root);
    while(!q.isEmpty()){
      int size=q.size();
      int max=q.peek().val;
      while(size-->0){
        TreeNode node = q.poll();
        max=max>node.val ? max : node.val;
        if(node.left!=null) q.offer(node.left);
        if(node.right!=null) q.offer(node.right);
      }
      res.add(max);
    }
    return res;
  }
}

File: 卡码/代码随想录/Go/二叉树/01-二叉树遍历方式/LeetCode_0515.go
//515. 在每个树行中找最大值
package main

import (
	. "binaryTree/types"
)

func largestValues(root *TreeNode) []int {
	res:=make([]int,0)
	q:=make([]*TreeNode,0)
	if root!=nil {
		q=append(q,root)
	}
	for len(q)>0 {
		size:=len(q)
		max:=q[0].Val
		for size>0 {
			size--
			node:=q[0]
			q=q[1:]
			if max<node.Val {
				max=node.Val
			}
			if node.Left!=nil {
				q=append(q,node.Left)
			}
			if node.Right!=nil {
				q=append(q,node.Right)
			}
		}
		res=append(res,max)
	}
	return res
}
